File: dotnet/niuz.application.test/services/ArticleServiceTest.cs
using niuz.application.entities;
using niuz.application.repositories;
using NSubstitute;
using Xunit;

namespace niuz.application.services
{
    public class ArticleServiceTest
    {
        private readonly IAuthorRepository authors;
        private readonly IArticleRepository articles;
        private readonly IPaymentRepository payments;
        private readonly ArticleService articleService;

        public ArticleServiceTest()
        {
            authors = Substitute.For<IAuthorRepository>();
            articles = Substitute.For<IArticleRepository>();
            payments = Substitute.For<IPaymentRepository>();
            articleService = new ArticleService(authors, articles, payments);
        }

        [Fact]
        public void SubmitArticleAndPay()
        {
            authors.GetByAuthorId("author-1").Returns(new Author("author-1", "Freddy Kruger", "123-4567-89", "pay-by-submission", 50));

            articleService.Submit("article-1", "author-1", "headline");
            
            articles.Received().Save(new Article("article-1", "author-1", "headline"));
            payments.Received().Save(new Payment(50, "123-4567-89", "Freddy Kruger", "headline"));
        }

        [Fact]
        public void SubmitArticleWithoutPay()
        {
            authors.GetByAuthorId("author-1").Returns(new Author("author-1", "Freddy Kruger", "123-4567-89", "pay-by-publication", 100));

            articleService.Submit("article-1", "author-1", "headline");
            
            articles.Received().Save(new Article("article-1", "author-1", "headline"));
            payments.DidNotReceive().Save(Arg.Any<Payment>());
        }
    }
}

File: java/src/test/java/com/dddeurope/niuz/services/ArticleServiceTest.java
package com.dddeurope.niuz.services;

import com.dddeurope.niuz.entities.Article;
import com.dddeurope.niuz.entities.Author;
import com.dddeurope.niuz.entities.Payment;
import com.dddeurope.niuz.repositories.ArticleRepository;
import com.dddeurope.niuz.repositories.AuthorRepository;
import com.dddeurope.niuz.repositories.PaymentRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.mockito.Mockito.*;

public class ArticleServiceTest {
    private AuthorRepository authors;
    private ArticleRepository articles;
    private ArticleService articleService;
    private PaymentRepository payments;

    @BeforeEach
    void setUp() {
        authors = mock(AuthorRepository.class);
        articles = mock(ArticleRepository.class);
        payments = mock(PaymentRepository.class);
        articleService = new ArticleService(authors, articles, payments);
    }

    @Test
    void submitArticleAndPay() {
        when(authors.getByAuthorId("author-1")).thenReturn(new Author("author-1", "Freddy Kruger", "123-4567-89", "pay-by-submission", 50));

        articleService.submit("article-1", "author-1", "headline");

        verify(articles).save(new Article("article-1", "author-1", "headline"));
        verify(payments).save(new Payment(50, "123-4567-89", "Freddy Kruger", "headline"));
    }

    @Test
    void submitArticleWithoutPay() {
        when(authors.getByAuthorId("author-1")).thenReturn(new Author("author-1", "Freddy Kruger", "123-4567-89", "pay-by-publication", 100));

        articleService.submit("article-1", "author-1", "headline");

        verify(articles).save(new Article("article-1", "author-1", "headline"));
        verify(payments, never()).save(any());
    }
}
